### ISSUE TITLE
kube-proxy pod fails to run due to unmounted volumes

### ISSUE THREAD

**Issue by xhejtman** [rancher/rke2#4864]
**Environmental Info:**
RKE2 Version:  1.26.9+rke2r1

Node(s) CPU architecture, OS, and Version:  
Linux kub-b9.priv.cerit-sc.cz 6.2.0-33-generic #33~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Sep  7 10:33:52 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:
5 servers, 30 agetns

**Describe the bug:**
Sometimes it happens that `kube-proxy` is unable to start. I can see that sandbox should be present:
```
root@kub-b9:~# CONTAINER_RUNTIME_ENDPOINT="unix:///var/run/k3s/containerd/containerd.sock" crictl pods -o yaml --label io.kubernetes.pod.name=kube-proxy-kub-b9.priv.cerit-sc.cz
items:
- annotations:
    kubernetes.io/config.hash: ce051fc91f9e463593a1d45efa60be52
    kubernetes.io/config.seen: "2023-10-03T14:17:50.457343231+02:00"
    kubernetes.io/config.source: file
  createdAt: "1696336132298880783"
  id: e505595e4af3ccb8e46c9dda0a79ce997b89c33529600708a389b7874b386ad3
  labels:
    component: kube-proxy
    io.kubernetes.pod.name: kube-proxy-kub-b9.priv.cerit-sc.cz
    io.kubernetes.pod.namespace: kube-system
    io.kubernetes.pod.uid: ce051fc91f9e463593a1d45efa60be52
    tier: control-plane
  metadata:
    attempt: 1
    name: kube-proxy-kub-b9.priv.cerit-sc.cz
    namespace: kube-system
    uid: ce051fc91f9e463593a1d45efa60be52
  runtimeHandler: ""
  state: SANDBOX_READY

root@kub-b9:~#
```

However, it is not actually running as it cannot be stopped:
```
oot@kub-b9:~# CONTAINER_RUNTIME_ENDPOINT="unix:///var/run/k3s/containerd/containerd.sock" crictl stop e505595e4af3c
E1005 20:28:29.770933 1411714 remote_runtime.go:349] "StopContainer from runtime service failed" err="rpc error: code = NotFound desc = an error occurred when try to find container \"e505595e4af3c\": not found" containerID="e505595e4af3c"
FATA[0000] stopping the container "e505595e4af3c": rpc error: code = NotFound desc = an error occurred when try to find container "e505595e4af3c": not found
root@kub-b9:~#
```

It can be force removed:
```
root@kub-b9:~# CONTAINER_RUNTIME_ENDPOINT="unix:///var/run/k3s/containerd/containerd.sock" crictl rmp -f e505595e4af3c
Stopped sandbox e505595e4af3c
Removed sandbox e505595e4af3c
root@kub-b9:~#
```

However, it does not run again. `service rke2-agent restart` does not spawn the `kube-proxy` pod again.

```
root@kub-b9:~# CONTAINER_RUNTIME_ENDPOINT="unix:///var/run/k3s/containerd/containerd.sock" crictl pods -o yaml --label io.kubernetes.pod.name=kube-proxy-kub-b9.priv.cerit-sc.cz
items: []

root@kub-b9:~#
```

```
root@kub-b9:~# CONTAINER_RUNTIME_ENDPOINT="unix:///var/run/k3s/containerd/containerd.sock" crictl ps -a | grep kube-proxy
root@kub-b9:~#
```

kubelet log is attached.

[kubelet.log.gz](https://github.com/rancher/rke2/files/12850826/kubelet.log.gz)

**Comment by brandond** [comment]
> However, it is not actually running as it cannot be stopped:

You're correctly using `crictl pods` and `crictl rmp` to list and delete the pod, but `crictl stop` takes a container ID, not a pod sandbox ID. Are you looking for `crictl stopp`?

The kubelet log indicates that it's unable to complete a syncronization of pods, either from the apiserver, or the static pod manifests. Note that the seenSources map is empty:
```
I1005 20:37:54.212933 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:54.312994 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:54.412643 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:54.512747 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:54.613093 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:54.712862 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:54.813317 1424921 generic.go:222] "GenericPLEG: Relisting"
I1005 20:37:54.813374 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:54.913133 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:55.012751 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:55.112466 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:55.213111 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:55.312780 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:55.413237 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
E1005 20:37:55.468783 1424921 kubelet.go:2137] "Skipping pod synchronization" err="container runtime status check may not have completed yet"
I1005 20:37:55.516255 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:55.612449 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:55.712897 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:55.813073 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:55.817338 1424921 generic.go:222] "GenericPLEG: Relisting"
I1005 20:37:55.912480 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.013898 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.112502 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.212781 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.313332 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.412827 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.512568 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.613159 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.712676 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.813121 1424921 config.go:105] "Looking for sources, have seen" sources=[api file] seenSources=map[]
I1005 20:37:56.824273 1424921 generic.go:222] "GenericPLEG: Relisting"
```

It doesn't ever seem to complete synchronizing the pod sources.

Are you able to reproduce this if you don't change the cpu manager policy, but just run with the default? I see it changing several times in the logs. 

Can you provide an updated copy of your config.yaml, as well as your containerd.config.toml if you're customizing the containerd config?

**Comment by xhejtman** [comment]
I do not have longhorn installed.

**Comment by brandond** [comment]
I have been doing some tracing of this issue, and it appears that there is a race condition in the kubelet startup, where it will fail to start a static pod if the static pod's mirror pod is discovered from the apiserver at the same time volumes are being mounted. I don't have a good fix for it at the moment, other than removing the kube-proxy static pod from disk prior to starting rke2 - this seems to delay the static pod creation long enough to avoid triggering the race condition. Unfortunately I haven't been able to reproduce the issue myself, so I am relying on some teammates to provide logs from an environment where it does seem to reoccur intermittently.

**Comment by xhejtman** [comment]
@brandond as one of our nodes is still in this state - kube-proxy does not start - I can try/trace anything needed here (I could provide you even remote ssh access).

**Comment by brandond** [comment]
Can you try deleting `/var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml`, and then restarting the service?

**Comment by xhejtman** [comment]
Yes, this helps on a different node, but not on kub-b9, I noticed this, that if I removed the manifest, kube proxy starts fine on other nodes, I thought it is because a new UID is generated..

Did you mean 'remove from disk' removing content from containerd or the abovementioned manifest? I could try the first, but if it fixes the node, we lose the testing case..

**Comment by brandond** [comment]
Just do exactly as I said and delete the file. If that does not resolve it, you may try also running `rke2-killall.sh` after deleting the file, although you should note that this will also terminate any other pods on this node.

**Comment by xhejtman** [comment]
```
root@kub-b9:~# rm /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml
root@kub-b9:~# service rke2-agent restart
root@kub-b9:~# ps fax | grep kube-proxy
2483805 pts/0    S+     0:00          \_ grep --color=auto kube-proxy
   8871 ?        Ssl    1:16  \_ pushprox-client --fqdn=10.16.62.44 --proxy-url=http://pushprox-kube-proxy-proxy.cattle-monitoring-system.svc:8080 --allow-port=10249 --use-localhost
2255256 ?        Ssl    2:08  \_ pushprox-client --fqdn=10.16.62.44 --proxy-url=http://pushprox-kube-proxy-proxy.cattle-monitoring-system.svc:8080 --allow-port=10249 --use-localhost
root@kub-b9:~#
```

even after few minutes, `kube-proxy` did not start. 

after `rke2-killall.sh` and `service rke2-agent start`, the `kube-proxy` finally started.

**Comment by brandond** [comment]
> Im pretty sure that if your event-config is setup to be super low and you have a daemonset failing rapidly this happens by itself after 24h. 

I'm sorry, you're saying that for some reason the kube-proxy static pod wasn't running because of an issue with an unrelated daemonset? Or are you talking about one of your daemonset pods not running?

**Comment by TravDev01** [comment]
Hey,
this also happens on a fresh install, at least for me.
Deleting the file was not working.
Let me know if any log files will help you.

I'm new to rancher and it was not a really smooth experience to have such issues.
Is there any stable version, where this problem does not exist?

**Comment by xhejtman** [comment]
> Hey, this also happens on a fresh install, at least for me. Deleting the file was not working. Let me know if any log files will help you.
> 
> I'm new to rancher and it was not a really smooth experience to have such issues. Is there any stable version, where this problem does not exist?

what rke2 version do you use?

**Comment by TravDev01** [comment]
I'm using v1.26.9+rke2r1 on an ubuntu 22.04 cloud image.
Basically I'm just following the quick start guide and end up without a kube-proxy and the install-cni pod is failing because it could not connect to the apiserver.

**Comment by xhejtman** [comment]
> I'm using v1.26.9+rke2r1 on an ubuntu 22.04 cloud image. Basically I'm just following the quick start guide and end up without a kube-proxy and the install-cni pod is failing because it could not connect to the apiserver.

version 1.26.7+rke2r1 should work fine. But remove everything from `/var/lib/rancher/rke2/agent/containerd/` if this is fresh install. `install-cni` cannot connect because the `kube-proxy` is not running.

and perhaps you can remove the `/var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml` and run `rke2-killall.sh` and restart rke, this could fix start as well.

**Comment by brandond** [comment]
@TravDev01 I have never seen this issue on a fresh install. I have seen what LOOKS like a similar issue on nodes that do not have sufficient CPU or memory resources available - the control-plane fails to start completely, which blocks kube-proxy from running. Can you provide the rke2-server log from journald, as well as information on the core count and memory size on your node?

**Comment by TravDev01** [comment]
> @TravDev01 I have never seen this issue on a fresh install. I have seen what LOOKS like a similar issue on nodes that do not have sufficient CPU or memory resources available - the control-plane fails to start completely due to lack of resources, which blocks kube-proxy from running. Can you provide the rke2-server log from journald, as well as information on the core count and memory size on your node?

Oh wow.
You are absolutely right.

I double-checked my resource configuration and noticed that proxmox was resetting it when I restored my snapshot between my various tests.
With the correct configuration everything is working, also with the lastet stable version.

Might it be beneficial to have a warning message in the log during the first bootstrap, when a user tries to create a cluster with fewer than the recommended resources?
For me, it's an interesting behavior that the pod is not even deployed. Normally, when you have a resource problem, you will see the pod in Kubernetes, but it stays in a pending state, because there is no node that is capable of running that pod.

Thanks a lot!

**Comment by brandond** [comment]
I don't know they there is an issue tracking this upstream, but the kubelet doesn't give any feedback if static pods fail to schedule. You have to turn the log up to like v=4 or higher if I remember correctly, and then there is a message complaining about scheduling and resources.

**Comment by brandond** [comment]
> in the middle of the night i interestingly have a restart

The service was explicitly stopped, can you tell why that happened?

**Comment by stephanbertl** [comment]
We are experiencing the same issue, kube-proxy (v1.26.8-rke2r1-build20230824) stuck and only this will fix it (temporarily)

```
rm /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml
systemctl restart rke2-agent
```

Happens across different types of nodes, all running Ubuntu 22.04

Sometimes we see these events on the kube-proxy:
```
Events:
  Type     Reason          Age   From     Message
  ----     ------          ----  ----     -------
  Normal   SandboxChanged  25m   kubelet  Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled          25m   kubelet  Container image "docker.dev.sg.info/rancher/hardened-kubernetes:v1.25.13-rke2r1-build20230824" already present on machine
  Normal   Created         25m   kubelet  Created container kube-proxy
  Normal   Started         25m   kubelet  Started container kube-proxy
  Warning  FailedMount     22m   kubelet  Unable to attach or mount volumes: unmounted volumes=[file2 file3 file0 file1], unattached volumes=[file2 file3 file0 file1]: timed out waiting for the condition
  Normal   Killing         12m   kubelet  Stopping container kube-proxy
```

**Comment by brandond** [comment]
@stephanbertl can you provide information on anything that might be unique about your environment? Being able to reproduce this on demand would be very helpful in isolating the root cause.

What is going on prior to kube-proxy getting stuck? How many cpus / how much memory do these nodes have? Is this happening only on agents, or also on servers?

**Comment by stephanbertl** [comment]
@brandond I only saw this on agents. Our environment isn't that special, and it happens on different node types: VMs on ESXi, bare-metal workstations and bare-metal servers.

One thing I also noticed it that we upgraded rke2 from 1.25 to1.26 through rancher and not all kube-proxy instances got updated. I again had to manually issue this to get them up-to-date.

```
rm /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml
systemctl restart rke2-agent
```

Today I saw it again, I just rebooted the Ubuntu 22.04 node.

**Comment by FlorianFuss** [comment]
Hello all,
we recently had a very similar issue in one of our rke2 clusters. kube-proxy wasnt able to start on the worker nodes. Only after killing the mirror-pod "kubectl delete pod kube-proxy-node1 -n kube-system --force" and restarting the rke2-agent service we could resolve the issue. A reboot did not fix the issue reliably.
We are also using version v1.26.9+rke2r1 but on RHEL8 nodes. The issue also occurs with version v1.27.5+rke2r1
In a testing environment we were able to replicate the issue by doing:
- Executing "rke2-killall.sh"
- Restarting rke2-agent "systemctl restart rke2-agent"

Most of the time after these steps the kube-proxy could not start anymore although it worked before like normal.
The fix mentioned in the comment above
```
rm /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml
systemctl restart rke2-agent
```
Fixed the issue reliably.

From my understanding with the resolution of https://github.com/rancher/rke2/issues/3725 the kube-proxy often gets a new UID (.metadata.uid) in the static-pod manifest when the rke2-agent (re)starts.

```
[root@node1 agent]# grep uid pod-manifests/kube-proxy.yaml
uid: 525b1d93649a7e4e2ab1a8cfb4004ef6
[root@node1 agent]# systemctl restart rke2-agent.service
[root@node1 agent]# grep uid pod-manifests/kube-proxy.yaml
uid: 0c29425699c4a36ff1b141176df07c4d
```

Although this is not always the case.

I believe that the kubelet cant correctly handle the change of UID of the kube-proxy static-pod the way rke2-agent does it. A possible problem could be the mismatch of UID in the static-pod manifest and the UID of the mirror-pod in the kube-apiserver.
Additionally we found an oddity in the timings between kubelet starting + reading the kube-proxy static-pod manifest and the creation/modification of the static-pod manifest of the kube-proxy by the rke2-agent.
It seems like the kubelet starts and reads the static-pod manifest of kube-proxy with UID "XYZ" which starts the process of creating this pod. 100-200 milliseconds later the rke2-agent writes a new static-pod manifest for the kube-proxy with UID "ABC" but otherwise unchanged.
This seems to lead to a kind of deadlock in the kubelet. The pod that got started first (UID "XYZ") cant be deleted properly and the second pod (UID "ABC") cant be created.

kubelet.log:
`I1201 14:53:43.547268 1701971 common.go:73] "Generated pod name" pod="kube-system/kube-proxy-node1" podUID=2df003787c169a83ff48577ebf058b47 source="/var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml"`
...
`I1201 14:53:43.728401 1701971 common.go:73] "Generated pod name" pod="kube-system/kube-proxy-node1" podUID=525b1d93649a7e4e2ab1a8cfb4004ef6 source="/var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml"`

>$ stat pod-manifests/kube-proxy.yml
Access: 2023-12-01 14:53:43.726296371 +0100
Modify: 2023-12-01 14:53:43.725296364 +0100
Change: 2023-12-01 14:53:43.725296364 +0100
Birth: 2023-12-01 14:53:43.725296364 +0100

As you can see here, the kubelet starts a kube-proxy with UID "2df003787c169a83ff48577ebf058b47" but the static-pod manifest gets modified ~180ms later by the rke2-agent process with a new UID "525b1d93649a7e4e2ab1a8cfb4004ef6".

Following log-entry (kubelet.log) can be observed multiple times after this happens:
`I1201 14:54:29.787237 1701971 pod_workers.go:1265] "Pod worker has been requested for removal but is still not fully terminated" podUID=2df003787c169a83ff48577ebf058b47`

Also this log-entry (kubelet.log) appears multiple times:
`I1201 14:54:11.785335 1701971 pod_workers.go:896] "Pod cannot start yet" pod="kube-system/kube-proxy-node1" podUID=525b1d93649a7e4e2ab1a8cfb4004ef6`

Sadly i cant append the full kubelet.log here because it contains sensitive information about our environment.
We also opened a ticket in the SUSE customer support portal, so hopefully this issue can be resolved quickly aswell if its the same bug.

**Comment by brandond** [comment]
I am curious to see if the fix from https://github.com/kubernetes/kubernetes/pull/122211 resolves or improves this, as it related to handling mounts that are in an unknown state during startup.

**Comment by brandond** [comment]
> @FlorianFuss 
> From my understanding with the resolution of https://github.com/rancher/rke2/issues/3725 the kube-proxy often gets a new UID (.metadata.uid) in the static-pod manifest when the rke2-agent (re)starts.

It should only get a new UID when the configuration is changed, RKE2 is upgraded, or an etcd restore is done (on servers running etcd). The pod manifest should NOT change following a restart of the service or reboot of the node with no other change to state.

The UID change is necessary to update the mirror pod in the apiserver - if the UID does not change, the mirror pod displayed in the apiserver will not be updated. The kubelet SHOULD NOT be trying to run the pod based on the mirror pod spec, the proper desired state of the pod 100% comes from the static pod manifest, and the pod resource displayed in the apiserver is a read-only "mirror" of that.

**Comment by Tejeev** [comment]
So we've run into this in one of my customer's environments and my suggestion was to add a script via bootstrap that runs every few minutes (per cronjob) that will check if kube-proxy is running.  If not, it would just run the  config removal and agent restart:
```
rm /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml
systemctl restart rke2-agent
```

@jhoelzel I'm not sure I fully understand the problem, so please forgive me if this is a stupid suggestion, but could this work for you (or am I missing something obviously wrong with the idea that I should note to my customer?)

**Comment by stephanbertl** [comment]
To add some information: after updating RKE2 minor version from 1.26.8 to 1.26.11, all pods get updated except, kube-proxy pods on agent nodes (server node pods get updated). They stay in their old version. They also show this:

```
 Warning  FailedMount  2m19s  kubelet  Unable to attach or mount volumes: unmounted volumes=[file0 file1 file2 file3], unattached volumes=[file0 file1 file2 file3]: timed out waiting for the condition
```

The only fix is to manually remove the static pod manifest. Then the new image is used.

```
rm /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml
systemctl restart rke2-agent
```

**Comment by Martin-Weiss** [comment]
FYI - just ran into exactly this problem as well during the upgrade of RKE2 1.25.16 to 1.26.11.

This workaround helped for a moment:
```
rm /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml
systemctl restart rke2-agent
```
 -> in my case this issue did not show up on the "rke2-server" side of things - just on the "rke2-agent" side (all agents)..

**Comment by Tejeev** [comment]
So just an update, the addition of the simple bash script that checks if kube-proxy is running and if not, runs the config removal and agent restart successfully worked around the problem and I've not been made aware of any negative repercussions

**Comment by dfaltum** [comment]
Hi everyone!

I can reproduce the problem with v1.25.16+rke2r1 and v1.26.13+rke2r1. Workaround: rke2-killall and service restart.

**Comment by SebPlv** [comment]
Hi,

We are also facing the problem when upgrading from v1.25.14 to v1.26.14.
But we have over 30 clusters with thousands of workloads, doing a "rke2-killall and service restart" cannot be a solution. As we have statefull workloads we need to drain each node before restarting the service.
As we are using Rancher to deploy/upgrade our clusters, could there be a solution managed by Rancher during the upgrade ?

**Comment by VestigeJ** [comment]
Community please feel free to try and test these fixes (not in prod) as I am unable to try to reproduce the original errors. I can see though that the agent process is pulling and rewriting the kube-proxy.yaml on reboots but we already have this behavior on existing releases so it isn't net new.

Are you all able to divulge how many volumes you have mounted in these large clusters? Are you able to expand on how resource constrained the nodes may be? I'm trying to figure out where the bottle neck is exactly to reproduce but can't do so with my limited testing environments. 


// existing behavior - I basically have followed the steps associated with getting the various PVs in a cluster (trying to orphan them...) then intentionally wrecking the kube-proxy.yaml then immediately rebooting.

//after the reboot
$ sudo cat /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml
cat: /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml: No such file or directory
$ sudo cat /var/lib/rancher/rke2/agent/pod-manifests/kube-proxy.yaml
```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    component: kube-proxy
    tier: control-plane
  name: kube-proxy
  namespace: kube-system
  uid: 0237
spec:
  containers:
  - args:
    - --cluster-cidr=10.42.0.0/16
    - --conntrack-max-per-core=0
    - --conntrack-tcp-timeout-close-wait=0s
    - --conntrack-tcp-timeout-established=0s
    - --healthz-bind-address=127.0.0.1
    - --hostname-override=ip-1-1-1-54
    - --kubeconfig=/var/lib/rancher/rke2/agent/kubeproxy.kubeconfig
    - --proxy-mode=iptables
    command:
    - kube-proxy
    env:
    - name: FILE_HASH
      value: d12034e965e
    - name: NO_PROXY
      value: .svc,.cluster.local,10.42.0.0/16,10.43.0.0/16
    image: index.docker.io/rancher/hardened-kubernetes:v1.29.3-rke2r1-build20240315
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 8
      httpGet:
        host: localhost
        path: /healthz
        port: 10256
        scheme: HTTP
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 15
    name: kube-proxy
    resources:
      requests:
        cpu: 250m
        memory: 128Mi
    securityContext:
      privileged: true
    startupProbe:
      failureThreshold: 24
      httpGet:
        host: localhost
        path: /healthz
        port: 10256
        scheme: HTTP
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 5
    volumeMounts:
    - mountPath: /var/lib/rancher/rke2/agent/client-kube-proxy.crt
      name: file0
      readOnly: true
    - mountPath: /var/lib/rancher/rke2/agent/client-kube-proxy.key
      name: file1
      readOnly: true
    - mountPath: /var/lib/rancher/rke2/agent/kubeproxy.kubeconfig
      name: file2
      readOnly: true
    - mountPath: /var/lib/rancher/rke2/agent/server-ca.crt
      name: file3
      readOnly: true
  hostNetwork: true
  priorityClassName: system-cluster-critical
  volumes:
  - hostPath:
      path: /var/lib/rancher/rke2/agent/client-kube-proxy.crt
      type: File
    name: file0
  - hostPath:
      path: /var/lib/rancher/rke2/agent/client-kube-proxy.key
      type: File
    name: file1
  - hostPath:
      path: /var/lib/rancher/rke2/agent/kubeproxy.kubeconfig
      type: File
    name: file2
  - hostPath:
      path: /var/lib/rancher/rke2/agent/server-ca.crt
      type: File
    name: file3
status: {}
```

$ kg pv,pvc -o wide -A
```
NAME                             CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS      CLAIM               STORAGECLASS    VOLUMEATTRIBUTESCLASS   REASON   AGE     VOLUMEMODE
persistentvolume/block-pv        10Gi       RWO            Retain           Released    default/block-pvc                   <unset>                          5h9m    Block
persistentvolume/local-storage   1Gi        RWO            Recycle          Available                       local-storage   <unset>                          3h42m   Filesystem
persistentvolume/vol             3Gi        RWX            Retain           Available                       manual          <unset>                          3h43m   Filesystem
```

**Comment by rancher-max** [comment]
I'm going to close this out based on the fixes made and repro attempts, but we can reopen if it is still noticed after these releases: v1.29.4, v1.28.9, and v1.27.13.

**Comment by stephanbertl** [comment]
I upgraded my cluster to 1.27.14 and the issue is gone.

**Comment by BloodyIron** [comment]
> Those of you who can, just use cilium in kube-proxy-less mode. I didnt really have any luck with kube-proxy-less calico, but cillium worked right out of the box.

To add to the conversation, I have recently tried cilium in kube-proxy-less mode with a Rancher v2.7.6 environment (and all the rk2+other versions that results in), and I found Cilium extremely unstable. It would crash itself every single time (100% of the time) in a random time range between about 5mins and 1h30min. It would never fix itself, and to "fix" it would take extreme efforts, and I'm not even sure what the consistent steps to take would be. As such I actually highly recommend _against_ Cilium in kube-proxy-less mode as it seems to be unreliable code. I also want to add that there were zero consistencies in the logs whenever it happened, so I was completely unable to determine the root cause, let alone what to do about it.

Hence I've recently switched back to trying to use Calico + eBPF to solve my SourceIP problem. But I am having issues with that reaching the kubernetes api-server VIP, and the somewhat maybe working workaround is to use localhost/127.0.0.1 but then calico-kube-controllers never succeeds (despite everything else seeming to work). So I'm trying to figure out the solution to that.

I might just bump my Rancher install to v2.8.x and try what @stephanbertl mentioned and upgrade test cluster nodes to 1.27.14 (or higher). Because higher numbers are _always_ better, right?

Anyways... my 3 year struggle to solve SourceIP problems continues... I feel I'm almost at the end of it, but hoo boy is the actual Calico and Rancher RKE2 official documentation not producing the results they say they would in this regards!
